articles/qos_configurability.md

```diff
@@ -0,0 +1,379 @@
+
+#### Only declaring some parameters
+
+Instead of hidden parameters that correspond to the QoS policies, we could only declare parameters for the policies that the user want to be reconfigurable.
+That will lower the amount of declared parameters, but make publisher/subscription creation a bit more verbose.
+
+Example API:
+
+```cpp
+enum class QosPolicyKind {
+  Reliability,
+  ...,
+};
+
+/// Options that are passed in subscription/publisher constructor
+class QosOverridingOptions
+{
+    /// Constructor, allowing to declare all the "default" parameters. If `false` is passed, reconfiguring QoS is not allowed.
+    explicit QosOverridingOptions(bool declare_default_parameters, std::string id = "");
+    /// Implicit constructor, a parameter for each policy provided in the initializer list will be declared.
+    QosOverridingOptions(std::initializer_list<QosPolicyKind>, std::string id = "");
+    /// Same with a callback for accepting/rejecting the qos profile.
+    QosOverridingOptions(std::initializer_list<QosPolicyKind>, QosCallback callback, std::string id = "");
+    /// Declares the "default" parameters and pass a callback.
+    QosOverridingOptions(QosCallback callback, std::string id = "");
+};
+```
+
+```cpp
+// New method in Node class
+template<
+typename MessageT,
+typename AllocatorT = std::allocator<void>,
+typename PublisherT = rclcpp::Publisher<MessageT, AllocatorT>>
+std::shared_ptr<PublisherT>
+create_publisher(
+  const std::string & topic_name,
+  const rclcpp::QoS & qos,
+  const rclcpp::QosOverridingOptions & qos_options = QosOverridingOptions{false},
+  const PublisherOptionsWithAllocator<AllocatorT> & options = PublisherOptionsWithAllocator<AllocatorT>()
+);
+```
+
+```cpp
+// example usage
+node->create_publisher(
+    "my_topic",
+    user_qos_provided_in_code_1,
+    rclcpp::KeepLast(10),
+    {  // implicit, to lower verbosity
+        {QosPolicyKind::Reliability, QosPolicyKind::History, QosPolicyKind::HistoryDepth}, // only declare parameters for history depth, history kind, reliability
+        [](const QoSProfile qos) -> bool {
+            return qos.is_keep_all() || qos.history_depth() > 10u;  // constrains involving more than one QoS in callbacks
+        },
+        "my_id",  // id to disambiguate entities in same topic
+    });
+
+// nothing is reconfigurable
+node->create_publisher(
+    "my_topic",
+    user_qos_provided_in_code_2,
+    rclcpp::KeepLast(10));
+
+// allow reconfiguring durability
+node->create_publisher(
+    "another_topic",
+    user_qos_provided_in_code_3,
+    rclcpp::KeepLast(10),
+    QosOverridingOptions{QosPolicyKind::Durability});
+```
+
+The above API mentions having some "defaults". 
+The intend of that is to provide a less verbose option, in which case the user can opt-in qos reconfigurability by only opting-in a flag.
+
+Proposed defaults:
+- Reliability
+- History kind
+- History depth
+- Durability
+
+TBD: Do we want defaults?
```

Comment:
IMO, letting all the settings be reconfigurable is fine (if a user opts in), unless there's a strong reason to exclude some settings.

Reply:
Liveliness kind should be excluded, as `MANUAL_BY_TOPIC` requires special care by the node's author.
I don't have a particular concern about other QoS settings.

Reply:
Excluding liveliness kind by default sounds fine to me.

Reply:
Clarified in https://github.com/ros2/design/pull/300/commits/563ed642675bd1dbdefa14ac8b7950bcc1486ef1.